flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/config/PulsarSourceConfigUtils.java

```diff
@@ -234,25 +239,87 @@ public static void checkConfigurations(Configuration configuration) {
+        setConfig(config.getConsumerName(), builder::consumerName);
+        setConfig(config.getPriorityLevel(), builder::priorityLevel);
+        setConfig(config.getMaxPendingChunkedMessage(), builder::maxPendingChunkedMessage);
+        setConfig(
+                config.isAutoAckOldestChunkedMessageOnQueueFull(),
+                builder::autoAckOldestChunkedMessageOnQueueFull);
+        setConfig(config.getProperties(), builder::properties);
+        setConfig(
+                config.getMaxTotalReceiverQueueSizeAcrossPartitions(),
+                builder::maxTotalReceiverQueueSizeAcrossPartitions);
+        setConfig(config.isReadCompacted(), builder::readCompacted);
+        setConfig(config.getPatternAutoDiscoveryPeriod(), builder::patternAutoDiscoveryPeriod);
+        setConfig(config.getPatternAutoDiscoveryPeriod(), builder::patternAutoDiscoveryPeriod);
+        setConfig(config.getSubscriptionInitialPosition(), builder::subscriptionInitialPosition);
+        setConfig(config.getRegexSubscriptionMode(), builder::subscriptionTopicsMode);
+        setConfig(config.isReplicateSubscriptionState(), builder::replicateSubscriptionState);
+        setConfig(config.getDeadLetterPolicy(), builder::deadLetterPolicy);
+        setConfig(config.isAutoUpdatePartitions(), builder::autoUpdatePartitions);
+        setConfig(
+                config.getAutoUpdatePartitionsIntervalSeconds(),
+                seconds ->
+                        builder.autoUpdatePartitionsInterval(
+                                Math.toIntExact(seconds), TimeUnit.SECONDS));
+        if (config.isResetIncludeHead()) {
+            builder.startMessageIdInclusive();
         }
+        setConfig(config.getBatchReceivePolicy(), builder::batchReceivePolicy);
+        setConfig(config.getKeySharedPolicy(), builder::keySharedPolicy);
+        setConfig(config.isRetryEnable(), builder::enableRetry);
+        setConfig(config.isBatchIndexAckEnabled(), builder::enableBatchIndexAcknowledgment);
+        setConfig(
+                config.getExpireTimeOfIncompleteChunkedMessageMillis(),
+                millis ->
+                        builder.expireTimeOfIncompleteChunkedMessage(
+                                millis, TimeUnit.MILLISECONDS));
+        setConfig(config.isPoolMessages(), builder::poolMessages);
 
-        return consumerBuilder.subscribe();
+        return builder.subscribe();
+    }
+
+    private static <T> void setConfig(T value, java.util.function.Consumer<T> consumer) {
+        if (value != null) {
+            if (value instanceof Collection) {
+                if (!((Collection<?>) value).isEmpty()) {
+                    consumer.accept(value);
+                }
+            } else if (value instanceof Map) {
+                if (!((Map<?, ?>) value).isEmpty()) {
+                    consumer.accept(value);
+                }
+            } else {
+                consumer.accept(value);
+            }
```

Comment:
I don't understand why this change is needed at all.
Why isn't `ConsumerBuilder<T> consumerBuilder = client.newConsumer(schema).loadConf(configMap(config));` working anymore?

Again would it be easier to remove `ConsumerConfigurationData` and just use `Configuration`?

Reply:
> We still have the opportunity to change API until the final release. Would it help to get rid of `ClientConfigurationData`? It feels like there is quite a bit friction involved. And since you replicated all options to `ConfigOption`s, it would even be better to go `Configuration` all the way for Flink users.

Drop the `ClientConfigurationData` and `ConsumerConfigurationData` would be cool. I think we could just use the  `Configuration` all the way. Do we need a separate PR for this feature.

Reply:
I would be fine with a separate commit in this PR.

Reply:
Seems like we have to combine this mock server with config refactor in one PR. The tests on mock server need the removal of PulsarJsonUtils. I would divide them in two commits.